### PR TITLE
[REVIEW] Add: atomicAnd/Or/Xor 

### LINKED ISSUE #1357: [BUG] Device atomics missing atomicOr and atomicAnd
**Describe the bug**
Recently merged PR #1194 added overloads for most CUDA device atomics for cuDF data types. But atomicOr and atomicAnd were missed -- these are important for valid bit mask manipulation in kernels.

We should add these. 

Also, I believe most of the atomics added missed unsigned data types. E.g. `uint8_t` (the underlying type of `gdf_valid_type`), `uint16_t`, etc.

Note that I don't think it's appropriate to add overloads for atomicOr and atomicAnd for non-integral types. I think we should just have int and uint 8, 16, 32, and 64.

### PR DESCRIPTION
**Pull request by kovaltan**
Implement  `atomicAnd`, `atomicOr`, `atomicXor` overload for all signed/unsigned integer types. 

The supported types are:
  singed/unsigned integer 8/16/32/64 bits
Cuda natively supports 32/64 bitwise logical operations.

Close #1357

Implement:
- [x] atomicAnd
- [x] atomicOr
- [x] atomicXor
- [x] test cases for all integers

### CODE REVIEW

File: cpp/src/utilities/device_atomics.cuh [hunk 1]
```diff
@@ -757,3 +781,215 @@ cudf::timestamp atomicCAS(cudf::timestamp* address, cudf::timestamp compare, cud
...
+    return cudf::genericAtomicOperation(address, val, cudf::DeviceAnd{});
+}
+
+/**
+ * @overload int64_t atomicAnd(int64_t* address, int64_t val)
+ */
+__forceinline__ __device__
+int64_t atomicAnd(int64_t* address, int64_t val)
+{
+    using T = long long int;
```
Comment (harrism) [inline cpp/src/utilities/device_atomics.cuh]: Why do we need overloads for int64_t and uint64_t if CUDA natively supports them as you say in the comment?

File: cpp/tests/utilities_tests/device_atomics_test.cu [hunk 2]
```diff
@@ -40,7 +41,7 @@ void gpu_atomic_test(T *result, T *data, size_t size)
         atomicAdd(&result[0], data[id]);
         atomicMin(&result[1], data[id]);
         atomicMax(&result[2], data[id]);
-        atomicAdd(&result[3], data[id]);
+        cudf::genericAtomicOperation(&result[3], data[id], cudf::DeviceSum{});
```
Comment (harrism) [inline cpp/tests/utilities_tests/device_atomics_test.cu]: Why this change? Also maybe explain in a comment why these four operations are applied to four locations of result?

File: cpp/tests/utilities_tests/device_atomics_test.cu [hunk 3]
```diff
@@ -76,6 +77,21 @@ void gpu_atomicCAS_test(T *result, T *data, size_t size)
...
+void gpu_atomic_bitwiseOp_test(T *result, T *data, size_t size)
+{
+    size_t id   = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t step = blockDim.x * gridDim.x;
+
+    for (; id < size; id += step) {
+        atomicAnd(&result[0], data[id]);
+        atomicOr(&result[1], data[id]);
+        atomicXor(&result[2], data[id]);
+        cudf::genericAtomicOperation(&result[3], data[id], cudf::DeviceAnd{});
```
Comment (harrism) [inline cpp/tests/utilities_tests/device_atomics_test.cu]: Why are there two different atomicAnd tests?

File: cpp/src/utilities/device_atomics.cuh (hunk 1 above)
Comment (kovaltan) [reply cpp/src/utilities/device_atomics.cuh]: nvcc compiler errors with without explicit int64_t, int64_t declaration.
It seems nvcc distincts int64_t and long long int.

File: cpp/tests/utilities_tests/device_atomics_test.cu (hunk 2 above)
Comment (kovaltan) [reply cpp/tests/utilities_tests/device_atomics_test.cu]: > Why this change?

I realized that there is no explicit cudf::genericAtomicOperation test case.
I've added it instead of `atomicAdd` which is already tested.

> Also maybe explain in a comment why these four operations are applied to four locations of result?

The reason for four locations of result is that the implementation for `int8_t` is different for each location and needs to be tested that Updating on different byte in same 32bit are occurred at same time.
I will add comment.
